### PR TITLE
[AArch64] Fix pairing different types of registers when computing CSRs.

### PR DESCRIPTION
**Pull request by nocchijiang**
If a function has odd number of same type of registers to save, and the calling convention also requires odd number of such type of CSRs, an FP register would be accidentally marked as saved when producePairRegisters returns true.

This patch also fixes the AArch64LowerHomogeneousPrologEpilog pass not handling AArch64::NoRegister; actually this pass must be fixed along with the register pairing so i can write a test for it.

### CODE REVIEW

**Comment by nocchijiang** [comment]
@TNorthover could you please have a look at the patch and add reviewers for me (i don't have commit access)?

@kyulee-com Trying to ping the original author of `AArch64LowerHomogeneousPrologEpilog` pass.

**Comment by kyulee-com** [comment]
Thanks for finding and fixing this issue!
Can you share some numbers on why this specialization is worth rather than bailing out since it's actually trying to handle the case where CSRs are not pairable (although the helper function names/checks are named after a pair)?
I also have a few questions on this approach:
 - How is CFI unwind info generated? Is it still valid? I'm not sure how it works with the compact unwind where CSRs are expects to be paired in order (without skipping one).
 - I saw `xzr` is handled at last after skipping a missing CSR. However, saving/restoring `xzr` is unnecessary, strictly speaking. Should we align the helper name that refelects `xzr` (although it's no-op) or should we use a single load/store as opposed to a pair load/store for the last element?

**Comment by kyulee-com** [comment]
> This patch is not changing the behavior of unwind info generation.

Intuitively, this doesn't seem right as you also save and restore `xzr` at the bottom of stack.

If you compare your test before/after `-homogeneous-prolog-epilog`, it seems different codegen semantically.
For instance, it didn't save `x29` originally while it now saves `x29`. As mentioned above, the bottom of stack seems different. Originally it had `x28/x27` at `sp+16` while it is now with `x28/xzr` at `sp+0`. So, the subsequent instruction `str     x21, [sp, #8] ` in the function body seems overlapped with the latter.

I suggest adding a real test case to confirm the change. Also suggest adding more comments as the underlying assumption was changed quite a bit. In fact, I prefer to have a right implementation for easier maintenance/understanding although it might lead to more substantial changes.

Another note/tip on debugging. Although the original implementation does not have a flag to expand the helper in place, I think the transformation should be still semantically valid even if `shouldUseFrameHelper` were false always, although I don't expect the same code sequence as the original code.

**Comment by nocchijiang** [comment]
> For instance, it didn't save x29 originally while it now saves x29

I think this is the expected behavior. When `producePairRegisters` returns `true`, as long as `LR` is saved, `FP` will be unconditionally saved no matter if it is really needed (`AArch64FrameLowering::hasFP`). The same behavior can be observed on Darwin platform when switching off the homogeneous prolog epilog pass:

Sample IR:
```llvm
declare void @bar(i32 %i)

define void @test(i32 %i) nounwind minsize {
  call void asm sideeffect "mov x0, #42", "~{x0},~{x19},~{x20},~{x21},~{x22},~{x23},~{x24},~{x25},~{x26},~{x27}"() nounwind
  call void @bar(i32 %i)
  ret void
}
```

Output:
```assembly
	.section	__TEXT,__text,regular,pure_instructions
	.ios_version_min 7, 0
	.globl	_test                           ; -- Begin function test
	.p2align	2
_test:                                  ; @test
; %bb.0:
	stp	x28, x27, [sp, #-96]!           ; 16-byte Folded Spill
	stp	x26, x25, [sp, #16]             ; 16-byte Folded Spill
	mov	w8, w0
	stp	x24, x23, [sp, #32]             ; 16-byte Folded Spill
	stp	x22, x21, [sp, #48]             ; 16-byte Folded Spill
	stp	x20, x19, [sp, #64]             ; 16-byte Folded Spill
	stp	x29, x30, [sp, #80]             ; 16-byte Folded Spill
	; InlineAsm Start
	mov	x0, #42                         ; =0x2a
	; InlineAsm End
	mov	w0, w8
	bl	_bar
	ldp	x29, x30, [sp, #80]             ; 16-byte Folded Reload
	ldp	x20, x19, [sp, #64]             ; 16-byte Folded Reload
	ldp	x22, x21, [sp, #48]             ; 16-byte Folded Reload
	ldp	x24, x23, [sp, #32]             ; 16-byte Folded Reload
	ldp	x26, x25, [sp, #16]             ; 16-byte Folded Reload
	ldp	x28, x27, [sp], #96             ; 16-byte Folded Reload
	ret
                                        ; -- End function
.subsections_via_symbols
```

But I noticed that `LR`/`FP` is repeatedly saved in the helper function on Linux target because the lowering pass was assuming that `LR` would always be at an even index (0-based) of the CSRs list, which is broken by the odd CSRs cases. I decide to bail for the "odd index" cases (I don't think Swift is that popular in Android world).

> So, the subsequent instruction str     x21, [sp, #8]  in the function body seems overlapped with the latter.

I believe that `str x21, [sp, #8]` writes `x21` into the empty slot (where `xzr` was "saved") so the overlap is benign. I debugged around related code and found that `computeFreeStackSlots` (in `PrologEpilogInserter.cpp`) tries to put frame objects into the empty slots created by uneven CSRs. Anyway I decided to implement `str`/`ldr` in the lowering pass instead of storing/loading `xzr`.

File: llvm/lib/Target/AArch64/AArch64FrameLowering.cpp [hunk 1]
```diff
@@ -3028,10 +3046,29 @@ void AArch64FrameLowering::determineCalleeSaves(MachineFunction &MF,
...
+      if (HasUnpairedGPR64)
+        PairedReg = CSRegs[i % 2 == 0 ? i - 1 : i + 1];
+      else
+        PairedReg = CSRegs[i ^ 1];
+    }
+
+    // If the function requires all the GP registers to save (SavedRegs),
+    // and there are an odd number of GP CSRs at the same time (CSRegs),
+    // PairedReg could be in a different register class from Reg, which would
+    // lead to an FPR (usually D8) accidentally being marked saved.
```
Comment (kyulee-com) [inline llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:3063]: an FPR -> a FPR

File: llvm/lib/Target/AArch64/AArch64LowerHomogeneousPrologEpilog.cpp [hunk 2]
```diff
@@ -195,46 +198,84 @@ static MachineFunction &createFrameHelperMachineFunction(Module *M,
...
+  if (IsPaired)
+    MIB.addReg(Reg2);
+  MIB.addReg(Reg1)
       .addReg(AArch64::SP)
       .addImm(Offset)
       .setMIFlag(MachineInstr::FrameSetup);
 }
 
 /// Emit a load-pair instruction for frame-destroy.
+/// If Reg2 is AArch64::NoRegister, emit LTR instead.
```
Comment (kyulee-com) [inline llvm/lib/Target/AArch64/AArch64LowerHomogeneousPrologEpilog.cpp:243]: LTR -> LDR

**Comment by kyulee-com** [comment]
Thanks for clearing this up! Overall it looks good to me module minor comments.

**Comment by nocchijiang** [comment]
Fixed the typos and a bad test expectation.

**Review by kyulee-com — APPROVED** [review]
LGTM

**Comment by nocchijiang** [comment]
@kyulee-com Could you please merge the PR for me since I don't have commit access? Or do you think it is necessary for someone else to review it?

**Pull request merged**